### DISCUSSION TITLE
Optional Map, required properties

### DISCUSSION THREAD

**Discussion by mauriceackel** [firecmsco/firecms#206] [Q&A]
Hi @fgatti675,

I've got a question regarding validation... Is it possible to have a schema like this:

```
button: {
    dataType: 'map',
    title: 'Action Button',
    description: 'Optional button that navigates to a page',
    properties: {
        text: {
            title: 'Button text',
            description: 'Text of the button in the award hint',
            dataType: 'string',
           validation: { required: true },
        },
    },
    validation: { required: false },
},
```

That is: The parent map (`button`) is optional but one of it's child properties (`text`) is required. This might seem strange at first. However, I think this is a quite common use case. A property might not exist on an object but when it does, it has to be of a certain type. The TypeScript equivalent would be something like:

```
interface Foo {
    text: string;
}

interface Bar {
    foofoo: string;
    barbar?: Foo;
}
```

On a technical side I have already started to look into this. It seems there are two things that currently prevent this behavior:

1. [Value initialization](https://github.com/Camberi/firecms/blob/master/src/core/utils.ts#L108)

This line makes it, that a map is always initialized as `key: {}`, even if none of its properties has a value. In comparison other keys that don't have a value are not included in the initialization. This could easily be solved though and I already have a patch in mind. **I also think this should definitely be solved in some way. However, it might require some additional UI elements to remove the property if it already exists on the element?!**

2. [Data validation](https://github.com/Camberi/firecms/blob/master/src/form/validation.ts#L111)

Because there is no `.default(null).nullable()`, the default yup behavior is to check for this property.

**SOME ISSUES**
If this was to be implemented, I see a view issues:

1. If a user enters a value by mistake, the property (e.g. `button`) will have been initialized, meaning that it is now required to enter the child properties. With the current UI, I guess there would be no way around this. There would need to be a button (e.g. a small "-") for map types and basically also array types that allow to set the value back to `null` if they should be excluded.
2. When I implemented a prototype it also prevented validation of legitimately required fields (i.e. other, non-optional map properties). This was because I simply added `.default(null).nullable()` [here](https://github.com/Camberi/firecms/blob/master/src/form/validation.ts#L111). To prevent this behavior, one could check if the property is required ands only if not, add the `.nullable()`.
3. The behavior might be achievable by other means, e.g. by using a builder function and validating the map property according to some logic. If non of the other children has a value, the `required: true` validation property could be removed from them.

Have you ever considered the topic explained above? What are your thoughts on it? Am I thinking too complicated and there is an easier solution? Should I create a PR with my prototype?

Cheers
Maurice

**Comment by mauriceackel** [comment]
I just noticed the `pickOnlySomeKeys` config option defined [here](https://firecms.co/docs/api/interfaces/MapFieldConfig/). However, this options seems to not be available on the schema level. Furthermore, this would be somewhat impractical in our use case as we only want to add/remove the keys that are optional in the first place and not any key.

**Comment by fgatti675** [comment]
Hi @mauriceackel 
I really can't think of an easier solution than the ones you are mentioning.
It is quite tricky as you mention with the current UI... We might need to include an additional checkbox to the `MapField` so the user can decide to set the value to null explicitly, and hide the rest of the fields in the map. 
Also I would only show that checkbox if the user has explicitly marked `required: false`, so we don't polute the UI too much
Then you will need to adapt the validation based on that, which I anticipate is not straight forward
Let me know if you need help!
